File: examples/ios/RunAnywhereAI/RunAnywhereAI/Features/Voice/TTSViewModel.swift
import Foundation
import RunAnywhere
import Combine
import os

// MARK: - TTS ViewModel

/// ViewModel for Text-to-Speech functionality
///
/// Uses `RunAnywhere.tts.speak()` - the SDK handles all audio playback internally.
@MainActor
class TTSViewModel: VoiceComponentViewModelBase {
    // MARK: - Component Identity

    override var component: RASDKComponent { .tts }
    override var eventCategory: RAEventCategory { .tts }
    override var modelCategory: RAModelCategory { .speechSynthesis }

    // MARK: - Published Properties

    // Speaking State
    @Published var isSpeaking = false

    /// True once a phrase has been spoken in this session. `tts.speak` returns
    /// no metrics, so this is all the UI can report about the last utterance.
    @Published private(set) var didSpeak = false

    // Voice Settings
    //
    // Speed is the whole control set, deliberately, and the same on all four
    // apps. A `pitch` property lived here with no slider behind it, which read
    // like a control that had gone missing rather than one that was never
    // offered; iOS is the source of truth for that decision, so the property is
    // gone and `TtsOptions` is passed speed only.
    @Published var speechRate: Double = 1.0

    // MARK: - Initialization

    init() {
        super.init(loggerCategory: "TTS")
    }

    /// Initialize the TTS view model
    /// This method is idempotent - calling it multiple times is safe
    func initialize() async {
        guard beginInitialization() else { return }

        logger.info("Initializing TTS view model")

        subscribeToSDKEvents()
        await checkInitialModelState()
    }

    // MARK: - Model Management

    /// Load a model from the unified model selection sheet
    func loadModelFromSelection(_ model: RAModelInfo) async {
        isSpeaking = true
        await loadModel(from: model)
        isSpeaking = false
    }

    // MARK: - Speaking

    /// Speak the given text aloud
    ///
    /// The SDK handles audio synthesis and playback internally.
    /// - Parameter text: The text to speak
    func speak(text: String) async {
        logger.info("Speaking: \(text.prefix(50))...")
        isSpeaking = true
        errorMessage = nil
        didRequestStop = false

        do {
            // `tts.speak` returns the moment playback *starts* — synthesis is
            // awaited, playout is not (see its doc comment: "let handle = try
            // await ...; await handle.waitForPlayout()"). Treating that return
            // as "finished" flipped `isSpeaking` false while the speaker was
            // still talking, which took the Stop control and the waveform off
            // screen for the whole utterance and left the status line claiming
            // the phrase had already been read. The handle is the only thing
            // that knows when the sound actually ends.
            let handle = try await RunAnywhere.tts.speak(
                text,
                options: TtsOptions(speed: Float(speechRate))
            )
            activeSpeech = handle
            await handle.waitForPlayout()
            activeSpeech = nil
            // A stopped utterance was never heard in full, so it must not leave
            // the screen offering to play it "again".
            if !handle.interrupted {
                didSpeak = true
            }
            logger.info("Speech playback finished")
        } catch {
            // A user- or teardown-initiated stop surfaces here as
            // `.playbackInterrupted`; that is expected, not a failure to report.
            if !didRequestStop {
                logger.error("Speech failed: \(error.localizedDescription)")
                errorMessage = "Speech failed: \(error.localizedDescription)"
            }
        }

        activeSpeech = nil
        isSpeaking = false
    }

    /// Set when playback is stopped intentionally (Stop button or teardown) so the
    /// `.playbackInterrupted` thrown back into `speak` isn't shown as an error.
    private var didRequestStop = false

    /// The utterance currently playing out, if any.
    private var activeSpeech: SpeechHandle?

    /// Stop current speech
    func stopSpeaking() async {
        logger.info("Stopping speech")
        didRequestStop = true
        // Interrupt this utterance through its own handle rather than the
        // deprecated whole-engine `tts.stop()`; it also resolves the
        // `waitForPlayout()` above, so `speak` unwinds instead of hanging.
        if let handle = activeSpeech {
            await handle.interrupt()
        } else {
            await RunAnywhere.tts.stop()
        }
        isSpeaking = false
    }

    // MARK: - Cleanup

    /// Clean up resources - call from view's onDisappear
    func cleanup() {
        // Stop any in-flight playback so speech doesn't keep playing after the
        // screen is dismissed (mirrors STT/VAD cleanup stopping capture).
        didRequestStop = true
        let closing = activeSpeech
        activeSpeech = nil
        Task {
            if let closing {
                await closing.interrupt()
            } else {
                await RunAnywhere.tts.stop()
            }
        }
        cleanupBase()
    }

    // MARK: - SDK Event Handling

    /// TTS surfaces the voice id directly as its display name rather than
    /// resolving it through the model catalog.
    override func applyLoadedModel(_ model: RAModelInfo) {
        selectedModelId = model.id
        selectedModelName = model.id
    }
}

File: examples/flutter/RunAnywhereAI/lib/features/voice/tts_view_model.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:runanywhere/runanywhere.dart' as sdk;

import 'package:runanywhere_ai/features/models/model_types.dart';
import 'package:runanywhere_ai/features/voice/voice_component_view_model_base.dart';

/// TTSMetadata (matching iOS TTSMetadata / RATTSSpeakResult surface).
class TTSMetadata {
  final double durationMs;
  final int audioSize;
  final int sampleRate;

  const TTSMetadata({
    required this.durationMs,
    required this.audioSize,
    required this.sampleRate,
  });
}

/// ViewModel for the Text-to-Speech view (mirrors iOS `TTSViewModel.swift`).
///
/// Uses the simplified `RunAnywhere.speak()` API — the SDK handles all audio
/// synthesis and playback internally. The view only renders state.
class TTSViewModel extends VoiceComponentViewModelBase {
  TTSViewModel();

  // SDK-owned playback streams (RunAnywhere.speak plays internally).
  StreamSubscription<bool>? _playingSubscription;
  StreamSubscription<double>? _progressSubscription;

  // --- Component identity -----------------------------------------------------

  @override
  ModelCategory get modelCategory =>
      ModelCategory.MODEL_CATEGORY_SPEECH_SYNTHESIS;

  // --- UI state ----------------------------------------------------------------

  bool isGenerating = false;
  bool isPlaying = false;
  double currentTime = 0.0;
  double duration = 0.0;
  double playbackProgress = 0.0;

  /// Speech rate multiplier (0.5x – 2.0x).
  double _speechRate = 1.0;
  double get speechRate => _speechRate;
  set speechRate(double value) {
    _speechRate = value;
    notify();
  }

  /// While removed from the UI, the backend still supports pitch (iOS parity).
  double pitch = 1.0;

  /// Whether the loaded voice is the platform system TTS.
  bool isSystemTTS = false;

  /// Metadata of the last generated audio, or null.
  TTSMetadata? metadata;

  // --- Initialization -------------------------------------------------------------

  /// Initialize the TTS view model — idempotent.
  Future<void> initialize() async {
    if (!beginInitialization()) return;

    debugPrint('Initializing TTS view model');
    _subscribeToPlayback();
    subscribeToSDKEvents();
    await checkInitialModelState();
  }

  void _subscribeToPlayback() {
    // Subscribe to the SDK's speak() playback state.
    _playingSubscription =
        sdk.RunAnywhere.tts.playbackState.listen((playing) {
      isPlaying = playing;
      if (playing) isGenerating = false;
      notify();
    });

    _progressSubscription =
        sdk.RunAnywhere.tts.playbackProgress.listen((progress) {
      playbackProgress = progress;
      currentTime = duration * progress;
      notify();
    });
  }

  // --- Model management ------------------------------------------------------------

  /// Load a voice from the unified model selection sheet.
  Future<void> loadModelFromSelection(ModelInfo model) async {
    isGenerating = true;
    notify();
    await loadModel(model);
    isGenerating = false;
    notify();
  }

  @override
  Future<void> performLoad(ModelInfo model) =>
      sdk.RunAnywhere.models.load(model.id);

  @override
  void applyLoadedModel(ModelInfo model) {
    super.applyLoadedModel(model);
    isSystemTTS =
        model.framework == LLMFramework.INFERENCE_FRAMEWORK_SYSTEM_TTS;
    notify();
  }

  // --- Speaking ----------------------------------------------------------------------

  /// Speak the given text aloud. One SDK entry point for every TTS path:
  /// speak() synthesizes AND plays (system voices and on-device models alike).
  Future<void> speak(String text) async {
    if (text.isEmpty) {
      errorMessage = 'Please enter text to speak';
      notify();
      return;
    }

    debugPrint('Speaking: ${text.length} chars');
    isGenerating = true;
    errorMessage = null;
    metadata = null;
    notify();

    try {
      if (!hasModelSelected) {
        throw StateError(
            'TTS component not loaded. Please load a TTS voice first.');
      }

      final options = sdk.TtsOptions(speed: _speechRate, pitch: pitch);
      // `speak()` returns a playback `SpeechHandle`, not audio bytes, so the
      // metadata card comes from a separate `synthesize()` call — the same
      // request commons would otherwise run once inside `speak()`.
      final audio = await sdk.RunAnywhere.tts.synthesize(text, options: options);
      duration = audio.durationMs / 1000.0;
      metadata = TTSMetadata(
        durationMs: audio.durationMs.toDouble(),
        audioSize: audio.data.length,
        sampleRate: audio.sampleRate,
      );
      debugPrint('Synthesized: ${audio.sampleRate} Hz, ${audio.durationMs}ms');

      final handle = sdk.RunAnywhere.tts.speak(text, options: options);
      await handle.waitForPlayout();
      final speakError = handle.error;
      if (speakError != null) throw speakError;
    } catch (e) {
      debugPrint('Speech generation failed: $e');
      errorMessage = 'Speech generation failed: $e';
    }

    isGenerating = false;
    notify();
  }

  /// Stop current speech playback.
  Future<void> stopSpeaking() async {
    debugPrint('Stopping speech');
    await sdk.RunAnywhere.tts.stop();
  }

  // --- Cleanup ---------------------------------------------------------------------------

  @override
  void dispose() {
    unawaited(_playingSubscription?.cancel());
    unawaited(_progressSubscription?.cancel());
    super.dispose();
  }
}
